### DISCUSSION TITLE
MSVC tool simplification/modern versions only

### DISCUSSION THREAD

**Discussion by bdbaddog** [SCons/scons#4565] [Ideas]
Move discussion here:

> > So we could support >=10.0 with only vswhere in a new msvc tool, and just leave the old one for legacy, and that would simplify things?
> 
> I'm not sure how much it simplifies things yet. I am not clear on what exactly new vs legacy means from a source organization perspective and for supported features moving forward. I'm not intentionally being obtuse, I just don't have a clear internal picture of what the desired end goal looks like.
> 
> In many ways, versions 2017 and later are more complex than earlier versions. While the older versions can be pruned, I'm not sure what the magnitude of the effects are with regards to code size, complexity, or run-time performance.
> 
> _**In the near term, it might be possible to emulate the effects of pruning versions by adding a "minimum version" for detection and support without actually removing any code.**_
> 
> Detection of installation roots (either via vswhere or the registry) is reasonably straightforward. Versions of 10.0 to 14.0 detected with vswhere would still need to be "classified" (development vs express, etc) as vswhere provides no information other than the vs root which is similar to the registry queries. VS2015 is in general more challenging than other versions if one is interested in validation of arguments.
> 
> Vswhere _does_ appear to correctly detect the legacy build tools (14.0) and express versions (10.0-14.0) and _does not_ detect legacy SDK-only installations (a good thing).
> 
> The tradeoff between using vswhere vs the registry for 10.0-14.0 might be just be issuing fewer registry queries. The complexity may be largely the same as what is present now.
> 
> In any case, I'm happy to work with you both.

The goal is not to stop supporting old versions, but rather to stop working on them as they are unlikley to change at this point.
Then use develop for relatively (yeah I know) modern msvc's.
This should yield two benefits:
1 - The number of VMs and testing needed to be complete on new changes should shrink significantly
2 - Hopefully simplify the implementation if we know we can grab the info from vswhere.

**Comment by mwichmann** [comment]
Remember we also have #4392

**Reply by bdbaddog** [reply]
Your memory is better than mine here... ;)

**Comment by mwichmann** [comment]
That's because it's a discussion I initiated...

**Comment by jcbrill** [comment]
Post release, I would like to:
- Rework (move, rename, refactor) existing code without necessarily adding new functionality.
- Implement an internal msvc data structure of detected msvc instances/toolsets/tools.
- Implement full support for preview versions (Preview, Release, Any).
- Implement support for msvs edition types (Enterprise, Professional, Community, BuildTools, Express, etc).
- Eliminate the independent handling of the msvc exists and msvc setup code paths.
- Construct the internal data structure during detection and query the internal data structure during selection.

The most difficult problem is reworking the argument lists of existing functions based on a structure instance rather than a loose collection of variables.  Having done it once, perhaps a piecemeal approach can be adopted so no one PR is too large.

PR #4409 which will be closed soon has some recoverable local code and provided an insight into what is necessary.

It might make sense to do most of these items prior to the new/old split.  Reworking to a unified internal data structure may make it easier to split later.  Not sure.

**Reply by bdbaddog** [reply]
Did the merged PR have equivalent (or better) functionality to #4409 ?

**Comment by mwichmann** [comment]
Let's wait for the definitive answer, but there was a comment early on that suggested it was intended to replace it:

> It is my intention to abandon #4409 as it is unlikely to be adopted in any reasonable time frame due to the quantity of uncommitted local changes and the number of updates to the main branch while the branch was under development.

**Comment by jcbrill** [comment]
> Did the merged PR have equivalent (or better) functionality to https://github.com/SCons/scons/pull/4409 ?

Yes and no. The merged PR stayed true to the *original* intent of PR #4409.

The code state prior to attempting to add msvc preview versions in PR #4409 was roughly the starting point for the merged PR.

The following was added to the merged PR that was not in #4409:
- VSWHERE/vswhere.exe support and issue fixes
- Initial support for msvc 14.4X toolsets

#4409 reverted to WIP in status due to wandering into the weeds attempting to implement preview versions.  A mistake in hindsight as the full integration likely was not complete nor acceptable due to a significant amount of new code.  A few smaller PRs were adopted from the work but the new code to support 4409 was larger than that of the merged PR.

The original intent of 4409 and the merged PR was to accurately detect all *currently supported* versions and editions of msvc.  In that sense, the merged PR is equivalent and better.

There should be a non-trivial amount of recoverable code from uncommitted local changes in 4409 that may be of use moving forward including implementing preview versions.  Some of the renaming/refactoring tasks may have already been done earlier in the local branch.
